### PR TITLE
Client unit tests

### PR DESCRIPTION
**Pull request by Tanvez**
Added unit tests to Client.upload, Client.create_manifest.
Left out functional tests for now since those will hit the MPS API and currently there is no way to delete manifests once they are created.

### CODE REVIEW

File: tests/unit/test_client.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+        client = test_client
+
+        images = [{"label": "Test Image", "filepath": image_path}]
+        assets = client.upload(images, s3_path="testing")
+
+        assert assets is not None
+        assert assets[0].filepath == image_path
+        
+    def test_client_fail_upload(self, test_images, boto_session, test_client):
```
Comment (jaguillette) [inline tests/unit/test_client.py:41]: I'm not sure what the failure state is here. I'm assuming that the upload will fail because the bucket was removed, but maybe a longer test name or docstring would be helpful here.

File: tests/unit/test_bucket.py [hunk 2]
```diff
@@ -37,4 +37,4 @@ def test_upload_directory(self, images_dir, boto_session):
 
     def test_fail_upload_directory(self, images_dir):
         with pytest.raises(ClientError):
-            assert upload_directory(images_dir, self.test_bucket_name, s3_path)
+            assert upload_directory(images_dir, self.test_bucket_name, s3_path)
```
Comment (jaguillette) [inline tests/unit/test_bucket.py:40]: Is this change intentional?

File: tests/unit/test_client.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+            boto_session.resource('s3').create_bucket(Bucket=self.bucket_name)
+            client = test_client
+            image_path = test_images["27.586.1-cm-2016-02-09.tif"]["filepath"]
+            images = [{"label": "Test Image", "filepath": image_path}]
+            assert client.create_manifest(
+                manifest_level_metadata=self.manifest_level_metadata,
+            )
+
+    def test_client_fail_create_manifest_missing_meta_data(
```
Comment (jaguillette) [inline tests/unit/test_client.py:92]: Very minor, but I'd call this `test_client_fail_create_manifest_missing_metadata`, removing the last underscore to match the args in the function.

**Review by jaguillette — APPROVED** [review]
Looks great! I added a couple of minor comments, but they're style preferences rather than substantive changes. Thanks!

File: src/IIIFingest/client.py [hunk 4]
```diff
@@ -45,6 +45,8 @@ def __init__(
         environment: str = "qa",
         jwt_creds=None,
         boto_session=None,
+        ingest_endpoint=None,
+        job_endpoint=None,
```
Comment (arthurian) [inline src/IIIFingest/client.py:49]: Currently, the `ingest_endpoint` and `job_endpoint` attributes are automatically configured in the constructor function based on the value of `environment`, but I could see a case where the user might want to manually provide those and override the defaults. Is that why they are being added to the constructor? The only reason I ask is that I don't see any other changes having to do with that or any tests.

Comment (arthurian) [inline src/IIIFingest/client.py:49]: Currently, the `ingest_endpoint` and `job_endpoint` attributes are automatically configured in the constructor function based on the value of `environment`, but I could see a case where the user might want to manually provide those and override the defaults. Is that why they are being added to the constructor? The only reason I ask is that I don't see any other changes having to do with that or any tests.

File: tests/conftest.py [hunk 5]
```diff
@@ -86,3 +90,29 @@ def boto_session():
...
+
+@pytest.fixture
+def test_client(boto_session):
+    private_key = "secret123file"
+    with tempfile.NamedTemporaryFile() as fp:
+        fp.write(private_key.encode('utf-8'))
+        fp.flush()
+        private_key_path = fp.name
+        issuer = ("atomeka_test",)
+        kid = ("atomekadefault_test",)
```
Comment (arthurian) [inline tests/conftest.py:103]: Did you mean to set the `issuer` and `kid` to tuples instead of strings?

File: tests/unit/test_client.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+from botocore.exceptions import ClientError
+import pytest
+
+from moto import mock_s3
+from dotenv import load_dotenv
+
+import re
+from IIIFingest.client import Client
+
+load_dotenv()
```
Comment (arthurian) [inline tests/unit/test_client.py:12]: FYI pytest has a plugin that can do this automatically: [pytest-dotenv](https://pypi.org/project/pytest-dotenv/)

File: tests/unit/test_client.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+        image_path = test_images["27.586.1-cm-2016-02-09.tif"]["filepath"]
+
+        client = test_client
+
+        images = [{"label": "Test Image", "filepath": image_path}]
+        assets = client.upload(images, s3_path="testing")
+
+        assert assets is not None
+        assert assets[0].filepath == image_path
```
Comment (arthurian) [inline tests/unit/test_client.py:39]: In addition to checking the asset `filepath`, you might also consider checking that the uploaded asset has an `asset_id` since that is created automatically as part of the upload process. If the image has an `id` property, it's supposed to be included in the `asset_id`, so you might consider checking both cases (with and without an image id).

File: tests/unit/test_client.py (hunk 1 above)
Comment (arthurian) [reply tests/unit/test_client.py:41]: I agree, it's not immediately clear what the expected failure state is. If this is expected to fail because the bucket does not exist, it might be a good idea to make the test name a little more descriptive.

File: tests/unit/test_client.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+                'filepath': '/Users/vtan/harvard-atg/lts-iiif-ingest-service/images/27.586.1-cm-2016-02-09.tif',
+                'format': 'image/tiff',
+            }
+        ]
+        """
+
+        manifest = client.create_manifest(
+            manifest_level_metadata=self.manifest_level_metadata, assets=assets
+        )
+        root_id = "https://nrs-dev.lib.harvard.edu/URN-3:TEST:"
```
Comment (arthurian) [inline tests/unit/test_client.py:72]: I'm not quite clear on where this value comes from. Is this the same as the `MPS_MANIFEST_BASE_URL` from the settings (formatted for the dev environment)?

File: tests/unit/test_client.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+        item = manifest["items"][0]
+        assert len(manifest["items"]) == 1
+        # assert matches_root_id is not None
+        assert item["height"] == 564 and item["width"] == 3600
```
Comment (arthurian) [inline tests/unit/test_client.py:76]: In addition to checking that the canvas item's dimensions are correct, you might also consider checking that the `id` matches the expected asset URL:

https://github.com/Harvard-ATG/lts-iiif-ingest-service/blob/d944d989e0703d55806abd7f191174b39599b3db/src/IIIFingest/client.py#L157

And  also that the `service` is configured by default to show the full/max size:

https://github.com/Harvard-ATG/lts-iiif-ingest-service/blob/d944d989e0703d55806abd7f191174b39599b3db/src/IIIFingest/client.py#L158

File: tests/conftest.py [hunk 10]
```diff
@@ -86,3 +90,29 @@ def boto_session():
         aws_session_token="testing",
         region_name='us-east-1',
     )
+
+
+@pytest.fixture
+def test_client(boto_session):
```
Comment (arthurian) [inline tests/conftest.py:96]: Nice!

**Review by arthurian — APPROVED** [review]
Overall looks great @Tanvez! Just added a couple of minor questions and minor suggestions.

File: tests/unit/test_client.py (hunk 1 above)
Comment (Tanvez) [reply tests/unit/test_client.py:41]: Will do!

File: tests/unit/test_bucket.py (hunk 2 above)
Comment (Tanvez) [reply tests/unit/test_bucket.py:40]: This should have been included in the main branch. I am not sure what changed vs the original code.

File: src/IIIFingest/client.py (hunk 4 above)
Comment (Tanvez) [reply src/IIIFingest/client.py:49]: Ahh yes, this was an artifact from when I was trying to run functional test as to provide a mock endpoint if needed. Will remove since we aren't running any functional client test at the moment.

File: tests/conftest.py (hunk 5 above)
Comment (Tanvez) [reply tests/conftest.py:103]: ahh I think I copied this code from another unit test. Tuples is definitely unnecessary for this! Will change to use strings.

File: tests/unit/test_client.py (hunk 8 above)
Comment (Tanvez) [reply tests/unit/test_client.py:72]: Yes, it is the same one as `MPS_MANIFEST_BASE_URL`. I hard coded it for now since I couldn't understand how to pass parameters to the `test_client` fixture. [Link to docs](https://docs.pytest.org/en/7.1.x/how-to/fixtures.html#fixture-parametrize)

Comment (arthurian) [reply tests/unit/test_client.py:72]: What about using the base URLs configured by the client? The client has an `asset_base_url` and `manifest_base_url`, and I'm not sure which one you'd need here, but if you assume the client is being constructed correctly (that should be a separate test to verify), then you could rely on those values for this particular test. That would make it a little more self-documenting as far as what's expected for the `manifest["id"]` and where it's coming from. Alternatively, you could import the `MPS_MANIFEST_BASE_URL` from `IIIFIngest.settings` and just format it manually like the client is doing.

Comment (Tanvez) [reply tests/unit/test_client.py:72]: Yes, that makes much more sense! I was looking to test the manually formatted `manifest_url` against the ones generated from`create_manifest` to see if they are correctly generated. So I think manually generating the manifest from `MPS_MANIFEST_BASE_URL` would be ideal for this particular test.

**Comment by arthurian** [comment]
@Tanvez Changes look great! 👍

**Review by jaguillette — APPROVED** [review]
Agreed, the changes look great!
